### PR TITLE
Augmentation Annealing: Linearly Decay Aug Strengths During Training

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Standard competition ML practice: apply full augmentation early in training for diversity, then reduce augmentation strength in the final phase for fine-tuning on clean data. The model first learns robust features from augmented samples, then refines on the true data distribution.

Currently, augmentation strengths are constant throughout training:
- `--aug_gap_stagger_sigma 0.02` (constant)
- `--aug_dsdf2_sigma 0.05` (constant)
- `--aug aoa_perturb` (constant perturbation)

Late in training (epochs 140-155), the cosine LR is near zero and the EMA is accumulating. At this point, augmentation noise may be HURTING rather than helping — the model is trying to fine-tune but sees noisy data.

**Proposed:** Linearly decay the multiplicative sigma values from 100% at epoch 0 to some fraction at the final epoch. Test two decay targets:
- Config A: decay to 50% (final σ_gs=0.01, σ_dsdf2=0.025)
- Config B: decay to 0% (final σ_gs=0, σ_dsdf2=0) — full annealing

**Expected impact:** p_tan -0.5% to -2%. The model gets both diversity (early) and precision (late).

### Step 1: Add augmentation annealing

In Config:
```python
aug_anneal: float = 1.0  # Final augmentation multiplier (1.0 = no annealing, 0.0 = full decay)
```

In argparse:
```python
parser.add_argument('--aug_anneal', type=float, default=1.0,
                    help='Final aug strength multiplier. Linear decay from 1.0 to this value over training.')
```

### Step 2: Compute annealing factor each epoch

At the start of each training epoch (inside the epoch loop):
```python
if cfg.aug_anneal < 1.0:
    # Linear decay from 1.0 at epoch 0 to cfg.aug_anneal at max_epoch
    _aug_scale = 1.0 - (1.0 - cfg.aug_anneal) * (epoch / cfg.cosine_T_max)
    _aug_scale = max(_aug_scale, cfg.aug_anneal)  # clamp at target
else:
    _aug_scale = 1.0
```

### Step 3: Apply scaling to sigma-based augmentations

In the augmentation block (around line 1542), scale the sigmas:
```python
# Gap/stagger aug
_gs_sigma = cfg.aug_gap_stagger_sigma * _aug_scale
# ... use _gs_sigma instead of cfg.aug_gap_stagger_sigma

# DSDF2 aug
_dsdf2_sigma = cfg.aug_dsdf2_sigma * _aug_scale
# ... use _dsdf2_sigma instead of cfg.aug_dsdf2_sigma
```

**Note:** The `aoa_perturb` augmentation may also benefit from annealing, but start with just the sigma-based augs.

### Step 4: Run experiments — 2 configs × 2 seeds = 4 runs

**Config A: Anneal to 50%** (σ decays from full to half)
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/aug-anneal50-s42" --wandb_group phase6/aug-annealing \
  --aug_anneal 0.5 --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "nezuko/aug-anneal50-s73")
```

**Config B: Anneal to 0%** (σ decays from full to zero)
```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent nezuko \
  --wandb_name "nezuko/aug-anneal0-s42" --wandb_group phase6/aug-annealing \
  --aug_anneal 0.0 --seed 42 \
  [... same baseline flags as above ...]

# Seed 73 (same, --seed 73, wandb_name "nezuko/aug-anneal0-s73")
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B |
|--------|------|------|--------|-------|------|-----|
| anneal→50% | 42 | | | | | |
| anneal→50% | 73 | | | | | |
| **anneal→50% avg** | — | | | | | |
| anneal→0% | 42 | | | | | |
| anneal→0% | 73 | | | | | |
| **anneal→0% avg** | — | | | | | |
| **Baseline (no anneal)** | — | 13.05 | 7.70 | 28.60 | 6.55 | d7l91p0x, j9btfx09 |

Also print `_aug_scale` at key epochs (0, 50, 100, 140, 155) to verify annealing is working.

## Baseline

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results — Augmentation Annealing

### Implementation
- Added `--aug_anneal` flag: linear decay of `aug_gap_stagger_sigma` and `aug_dsdf2_sigma` from 1.0 at epoch 0 to target value at `cosine_T_max`
- Verified annealing prints at key epochs (0, 50, 100, 140, 155)

### Annealing Verification
| Epoch | anneal→50% scale | anneal→0% scale |
|-------|-----------------|----------------|
| 0 | 1.0000 | 1.0000 |
| 50 | 0.8438 | 0.6875 |
| 100 | 0.6875 | 0.3750 |
| 140 | 0.5625 | 0.1250 |

### Results (Surface MAE — Pressure)

| Config | Seed | p_in | p_oodc | p_tan | p_re | W&B ID |
|--------|------|------|--------|-------|------|--------|
| anneal→50% | 42 | 12.878 | 7.838 | 28.443 | 6.341 | `7ofuolg3` |
| anneal→50% | 73 | 13.294 | 7.857 | 29.348 | 6.380 | `zt31115v` |
| **anneal→50% avg** | — | **13.086** | **7.848** | **28.896** | **6.361** | |
| anneal→0% | 42 | 12.739 | 7.993 | 29.278 | 6.533 | `ibywi5rr` |
| anneal→0% | 73 | 13.796 | 7.817 | 29.125 | 6.439 | `4jd564uq` |
| **anneal→0% avg** | — | **13.268** | **7.905** | **29.202** | **6.486** | |
| **Baseline** | — | **13.05** | **7.70** | **28.60** | **6.55** | d7l91p0x, j9btfx09 |

### Analysis

**Neither config beats baseline on the key metric p_tan** (28.90 vs 28.60 for anneal→50%, 29.20 vs 28.60 for anneal→0%).

**anneal→50% is the better of the two** — it outperforms anneal→0% on every metric, suggesting keeping some augmentation active is better than fully removing it.

**Interesting finding: p_re improves significantly with anneal→50%** (6.36 vs 6.55 baseline, -2.9%). This makes sense — the OOD Reynolds split benefits from cleaner late-training data, since the augmentations mostly target tandem geometry features (gap/stagger) rather than Reynolds number variation.

**Why p_tan regressed:** The tandem transfer split likely depends on augmentation diversity throughout training. Annealing reduces the gap/stagger perturbation that helps the model generalize to unseen tandem configurations. The baseline's constant augmentation may be better matched to this specific generalization challenge.

### Suggested Follow-ups
1. **Selective annealing**: Only anneal `aug_dsdf2_sigma` (geometry noise) while keeping `aug_gap_stagger_sigma` constant — this could preserve p_tan while still gaining the p_re benefit
2. **Cosine annealing schedule** instead of linear — more gradual decay early, sharper reduction late (matching the LR schedule)
3. **Anneal only the last 20% of training** (step function at epoch 128 rather than linear throughout) — this is closer to competition "fine-tune on clean data" practice

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2152 because ADVISOR: Closing — augmentation annealing did not beat baseline on the key metric p_tan.

**Results summary:**
- anneal→50% avg: p_tan=28.90 vs baseline 28.60 (+1.0%) — regression
- anneal→0% avg: p_tan=29.20 vs baseline 28.60 (+2.1%) — worse regression
- Only p_re improved (6.36 vs 6.55, -2.9%) with anneal→50%

**Analysis:** The tandem transfer split (p_tan) depends on sustained augmentation diversity throughout training. Annealing the gap/stagger sigma reduces the domain randomization that helps generalize to unseen tandem configurations. The constant augmentation in the baseline is better matched to this generalization challenge. The p_re improvement is a secondary effect of cleaner late-training data for Reynolds generalization, but isn't worth the p_tan regression.

**Insight preserved:** Late-training augmentation strength is important for p_tan — confirms that gap/stagger sigma=0.02 constant throughout training is the right schedule. Selective annealing of only dsdf2 (not gap/stagger) could be interesting but the expected gain is too small to justify further iterations.

Good analysis, nezuko — the follow-up suggestions showed strong experimental intuition..
